src/repr/binary_codec_intf.ml

```diff
@@ -0,0 +1,72 @@
+open Type_core
+open Staging
+
+type 'a encoder = 'a -> (string -> unit) -> unit
+type 'a decoder = string -> int -> int * 'a
+
+module type S = sig
+  type t
+
+  val encode : t encoder
+  val decode : t decoder
+end
+
+module type S_with_length = sig
+  type t
+
+  val encode : len -> t encoder staged
+  val decode : len -> t decoder staged
+end
+
+module type S1 = sig
+  type 'a t
+
+  val encode : 'a encoder -> 'a t encoder
+  val decode : 'a decoder -> 'a t decoder
+end
+
+module type S1_with_length = sig
+  type 'a t
+
+  val encode : len -> 'a encoder -> 'a t encoder staged
+  val decode : len -> 'a decoder -> 'a t decoder staged
+end
+
+module type S2 = sig
+  type ('a, 'b) t
+
+  val encode : 'a encoder -> 'b encoder -> ('a, 'b) t encoder
+  val decode : 'a decoder -> 'b decoder -> ('a, 'b) t decoder
+end
+
+module type S3 = sig
+  type ('a, 'b, 'c) t
+
+  val encode : 'a encoder -> 'b encoder -> 'c encoder -> ('a, 'b, 'c) t encoder
+  val decode : 'a decoder -> 'b decoder -> 'c decoder -> ('a, 'b, 'c) t decoder
+end
```

Comment:
Could we give these signatures more eloquent names like `S_singleton`, `S_pair` and `S_triple` so maybe one day we'll be able to have `S_variadic`

Reply:
I'm not sure about `S_pair` and `S_triple`, since I think they conflate _"`t` has two type parameters `'a` and `'b`"_ with _"`t` is a pair of two types `'a` and `'b`"_, but I'm not opposed to abstracting over arity in future (I had a PPX for this pattern at one point...).

Will merge now to unblock other PRs, but we can review this later as it doesn't leak.